cmd/skopeo/list_tags.go

```diff
@@ -0,0 +1,142 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/containers/image/v5/docker"
+	"github.com/containers/image/v5/transports"
+	"github.com/containers/image/v5/transports/alltransports"
+	"github.com/containers/image/v5/types"
+	"github.com/docker/distribution/reference"
+	"github.com/pkg/errors"
+	"github.com/urfave/cli"
+	"strings"
+
+	"io"
+)
+
+// tagListOutput is the output format of (skopeo list-tags), primarily so that we can format it with a simple json.MarshalIndent.
+type tagListOutput struct {
+	Repository string
+	Tags       []string
+}
+
+type tagsOptions struct {
+	global *globalOptions
+	image  *imageOptions
+}
+
+func tagsCmd(global *globalOptions) cli.Command {
+	sharedFlags, sharedOpts := sharedImageFlags()
+	imageFlags, imageOpts := dockerImageFlags(global, sharedOpts, "", "")
+
+	opts := tagsOptions{
+		global: global,
+		image:  imageOpts,
+	}
+
+	return cli.Command{
+		Name:  "list-tags",
+		Usage: "List tags in the transport/repository specified by the REPOSITORY-NAME",
+		Description: `
+	Return the list of tags from the transport/repository "REPOSITORY-NAME"
+	
+    Supported transports:
+	docker
+
+	See skopeo-list-tags(1) section "REPOSITORY NAMES" for the expected format
+	`,
+		ArgsUsage: "REPOSITORY-NAME",
+		Flags:     append(sharedFlags, imageFlags...),
+		Action:    commandAction(opts.run),
+	}
+}
+
+// Customized version of the alltransports.ParseImageName and docker.ParseReference that does not place a default tag in the reference
+// Would really love to not have this, but needed to enforce tag-less and digest-less names
+func parseDockerRepositoryReference(refString string) (types.ImageReference, error) {
+	parts := strings.SplitN(refString, ":", 2)
+	if len(parts) != 2 {
+		return nil, errors.Errorf(`Invalid image name "%s", expected colon-separated transport:reference`, refString)
+	}
+
+	transport := transports.Get(parts[0])
+	if transport == nil || transport.Name() != docker.Transport.Name() {
```

Comment:
`run` already did a check for the right transport; it shouldn’t be necessary to do that twice. Preferably, do the checks in this function only, so that they can be covered by the unit tests.

(`alltransports.ParseImageName`, with no manual string parsing here, would be esthetically preferable, but that would not allow rejecting tagged/digested input. I’m honestly undecided whether the extra check is worth it…)

Alternatively, just checking explicitly for the `docker.transport.Name()+"://"` prefix, and stripping exactly that prefix, would do just as many assumptions about the format, and be simpler.

Reply:
> `run` already did a check for the right transport; it shouldn’t be necessary to do that twice. Preferably, do the checks in this function only, so that they can be covered by the unit tests.
> 

Ack. Yes, this was just a function-level sanity check to ensure the input is valid. A simpler check as you outline below is better since I made the change to a transport-specific parser.

> (`alltransports.ParseImageName`, with no manual string parsing here, would be esthetically preferable, but that would not allow rejecting tagged/digested input. I’m honestly undecided whether the extra check is worth it…)

Agreed. I'm of the opinion that it is needed since 'list-tags docker://myimage:latest' could be interpreted to mean "show me the other tags that also point to this image digest", but if you guys feel that docs are a sufficient means of clarification then I can remove a bunch of this redundant checking code.

> 
> Alternatively, just checking explicitly for the `docker.transport.Name()+"://"` prefix, and stripping exactly that prefix, would do just as many assumptions about the format, and be simpler.

Agree as above. I'll make this change to simplify things and remove redundant code.

Reply:
OK, looking at the `run` code as the one responsible for branching to different transports: a simple check+strip of the `docker.transport.name()+"///"` prefix, failing with an `Internal error: Docker reference expected` (to make it clear that the end-user-visible error handling happens elsewhere) would work well.